modules/libcom/src/osi/osiFileName.c

```diff
@@ -0,0 +1,197 @@
+/*************************************************************************\
+* Copyright (c) 2021 Michael Davidsaver
+* SPDX-License-Identifier: EPICS
+* EPICS Base is distributed subject to a Software License Agreement found
+* in file LICENSE that is included with this distribution.
+\*************************************************************************/
+
+#include <stdarg.h>
+#include <string.h>
+#include <errno.h>
+
+#include <epicsAssert.h>
+#include <osiUnistd.h>
+#include <epicsString.h>
+#include "osiFileName.h"
+
+#if defined(_WIN32) || defined(__CYGWIN__)
+// cf. https://docs.microsoft.com/en-us/dotnet/standard/io/file-path-formats#canonicalize-separators
+#  define SEP(C) (C=='/' || C=='\\')
+#else
+#  define SEP(C) (C=='/')
+#endif
+
+static const char magicCWD[] = "<current_directory>";
+const char* const epicsPathJoinCurDir = magicCWD;
+
+static
+size_t epicsPathDF(char *path, size_t pathlen, int outfile)
+{
+    size_t pos;
+
+    /* search forward to ignore unused buffer */
+    pathlen = epicsStrnLen(path, pathlen);
+
+    /* search backwards for last separator */
+    pos = pathlen;
+    for(; pos>0; pos--) {
+        char c = path[pos-1u];
+        if(SEP(c)) {
+            /* found separator */
+            if(outfile) {
+                memcpy(path, &path[pos], pathlen-pos);
+                pos = pathlen-pos;
+            }
+            goto done;
+        }
+    }
+    /* no separator found */
+    if(outfile) {
+        pos = pathlen;
+    }
+
+done:
+    if(pos < pathlen) {
+        path[pos] = '\0';
+    }
+
+    return pos;
+}
+
+size_t epicsPathDir(char *path, size_t pathlen)
+{
+    return epicsPathDF(path, pathlen, 0);
+}
+
+size_t epicsPathFile(char *path, size_t pathlen)
+{
+    return epicsPathDF(path, pathlen, 1);
+}
+
+int epicsPathIsAbs(const char *path, size_t pathlen)
+{
+    /* for *NIX this is simple.  path[0]=='/' is absolute.  Anything else is relative
+     *
+     * for windows... not so much.
+     * cf. https://docs.microsoft.com/en-us/dotnet/standard/io/file-path-formats#identify-the-path
+     *
+     * we treat as absolute:
+     *  - device and UNC paths.
+     *  - absolute on current drive.  eg. "\foo" -> "C:\foo" if %CD% starts with C:
+     *  - absolute w/ drive.  eg. "C:\foo"
+     *
+     * we treat as relative
+     *  - relative to some drive.  eg. "C:foo" -> "D:\bar\foo" if %CD% == "D:\bar"  (sneaky!)
```

Comment:
It is still relative, you just have to be careful if you tried to reconstruct an absolute path from it yourself - it is relative to a directory that may not be your current working directory (i.e. `getdcwd` rather than `getcwd`)